app/src/main/java/com/quran/labs/androidquran/ui/fragment/JumpFragment.java

```diff
@@ -132,50 +125,47 @@ public void afterTextChanged(Editable s) {
         String ayahString = s.toString();
         int ayah = parseInt(ayahString, 1);
 
-        Object suraTag = suraSpinner.getTag();
+        Object suraTag = suraInput.getTag();
         if (suraTag != null) {
           int sura = (int) suraTag;
           int ayahCount = QuranInfo.getNumAyahs(sura);
           ayah = Math.max(1, Math.min(ayahCount, ayah)); // ensure in 1..ayahCount
           int page = QuranInfo.getPageFromSuraAyah(sura, ayah);
-          input.setHint(QuranUtils.getLocalizedNumber(context, page));
-          input.setText(null);
+          pageInput.setHint(QuranUtils.getLocalizedNumber(context, page));
+          pageInput.setText(null);
         }
 
-        ayahSpinner.setTag(ayah);
+        ayahInput.setTag(ayah);
         // seems numeric IM always use western arabic (not localized)
         String correctText = String.valueOf(ayah);
+        // empty input means the user clears the input, we don't force to fill it, let him type
         if (s.length() > 0 && !correctText.equals(ayahString)) {
           s.replace(0, s.length(), correctText);
         }
       }
     });
 
     builder.setView(layout);
-    builder.setPositiveButton(getString(R.string.dialog_ok), new DialogInterface.OnClickListener() {
-      @Override
-      public void onClick(DialogInterface dialog, int which) {
-        try {
-          dismiss();
-          String text = input.getText().toString();
-          if (TextUtils.isEmpty(text)) {
-            text = input.getHint().toString();
-            int page = Integer.parseInt(text);
-            int selectedSura = (int) suraSpinner.getTag();
-            int selectedAyah = (int) ayahSpinner.getTag();
-
-            Activity activity = getActivity();
-            if (activity instanceof QuranActivity) {
-              ((QuranActivity) activity).jumpToAndHighlight(page, selectedSura, selectedAyah);
-            } else if (activity instanceof PagerActivity) {
-              ((PagerActivity) activity).jumpToAndHighlight(page, selectedSura, selectedAyah);
-            }
-          } else {
-            goToPage(text);
+    builder.setPositiveButton(getString(R.string.dialog_ok), (dialog, which) -> {
+      try {
+        dismiss();
+        String pageStr = pageInput.getText().toString();
+        if (TextUtils.isEmpty(pageStr)) {
+          pageStr = pageInput.getHint().toString();
+          int page = Integer.parseInt(pageStr);
+          int selectedSura = (int) suraInput.getTag();
+          int selectedAyah = (int) ayahInput.getTag();
+
+          if (activity instanceof QuranActivity) {
```

Comment:
i know this is existing but we should really clean this up. implement an interface in both QuranActivity and PagerActivity and just be like if it's this interface call jumpAndHighlight